File: nestjs/src/auth/auth.service.ts
import { Injectable, UnauthorizedException, Logger } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { ConfigService } from '@nestjs/config';
import { UsersService } from '../users/users.service';
import { RegisterDto, LoginDto } from './dto/auth.dto';
import type { User } from '../types/prisma.types';

export interface JwtPayload {
  userId: string;
  iat?: number;
  exp?: number;
}

@Injectable()
export class AuthService {
  private readonly logger = new Logger(AuthService.name);

  constructor(
    private readonly usersService: UsersService,
    private readonly jwtService: JwtService,
    private readonly configService: ConfigService,
  ) {}

  async register(
    registerDto: RegisterDto,
  ): Promise<{ accessToken: string; refreshToken: string }> {
    const { email, password } = registerDto;

    const existingUser = await this.usersService.findByEmail(email);
    if (existingUser) {
      throw new UnauthorizedException('User already exists');
    }

    const user = await this.usersService.create(email, password);

    const accessToken = this.generateAccessToken(user.id);
    const refreshToken = this.generateRefreshToken(user.id);

    await this.usersService.addRefreshToken(user.id, refreshToken);

    return { accessToken, refreshToken };
  }

  async login(
    loginDto: LoginDto,
  ): Promise<{ accessToken: string; refreshToken: string }> {
    const { email, password } = loginDto;

    const user = await this.usersService.findByEmail(email);
    if (!user) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const isPasswordValid = await this.usersService.validatePassword(
      user,
      password,
    );

    if (!isPasswordValid) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const accessToken = this.generateAccessToken(user.id);
    const refreshToken = this.generateRefreshToken(user.id);

    await this.usersService.addRefreshToken(user.id, refreshToken);

    return { accessToken, refreshToken };
  }

  async refresh(
    oldRefreshToken: string,
  ): Promise<{ accessToken: string; refreshToken: string }> {
    let payload: JwtPayload;

    try {
      payload = this.jwtService.verify<JwtPayload>(oldRefreshToken, {
        secret: this.configService.getOrThrow<string>('REFRESH_TOKEN_SECRET'),
      });
    } catch (error) {
      throw new UnauthorizedException('Invalid refresh token : ' + error);
    }

    const user = await this.usersService.findById(payload.userId);
    if (!user) {
      throw new UnauthorizedException('User not found');
    }

    const hasToken = await this.usersService.hasRefreshToken(
      user.id,
      oldRefreshToken,
    );

    if (!hasToken) {
      throw new UnauthorizedException('Refresh token revoked');
    }

    const newAccessToken = this.generateAccessToken(user.id);
    const newRefreshToken = this.generateRefreshToken(user.id);

    await this.usersService.replaceRefreshToken(
      user.id,
      oldRefreshToken,
      newRefreshToken,
    );

    return { accessToken: newAccessToken, refreshToken: newRefreshToken };
  }

  async logout(userId: string, refreshToken: string): Promise<void> {
    try {
      await this.usersService.removeRefreshToken(userId, refreshToken);
    } catch (error: unknown) {
      if (error instanceof Error) {
        this.logger.warn(`Failed to remove refresh token: ${error.message}`);
      } else {
        this.logger.warn('Failed to remove refresh token: unknown error');
      }
    }
  }

  async validateUser(userId: string): Promise<User> {
    const user = await this.usersService.findById(userId);
    if (!user) {
      throw new UnauthorizedException('User not found');
    }
    return user;
  }

  private generateAccessToken(userId: string): string {
    return this.jwtService.sign(
      { userId },
      {
        secret: this.configService.getOrThrow<string>('ACCESS_TOKEN_SECRET'),
        expiresIn:
          this.configService.get<string>('ACCESS_TOKEN_EXPIRY') ?? '15m',
      },
    );
  }

  private generateRefreshToken(userId: string): string {
    return this.jwtService.sign(
      { userId },
      {
        secret: this.configService.getOrThrow<string>('REFRESH_TOKEN_SECRET'),
        expiresIn:
          this.configService.get<string>('REFRESH_TOKEN_EXPIRY') ?? '7d',
      },
    );
  }
}

File: springboot/src/main/java/com/example/springboot/auth/AuthService.java
package com.example.springboot.auth;

import com.example.springboot.auth.dto.LoginRequest;
import com.example.springboot.auth.dto.RegisterRequest;
import com.example.springboot.user.User;
import com.example.springboot.user.UserRepository;
import java.util.ArrayList;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.http.HttpStatus;

@Service
public class AuthService {
  private final UserRepository userRepository;
  private final PasswordEncoder passwordEncoder;
  private final JwtService jwtService;

  public AuthService(UserRepository userRepository, PasswordEncoder passwordEncoder, JwtService jwtService) {
    this.userRepository = userRepository;
    this.passwordEncoder = passwordEncoder;
    this.jwtService = jwtService;
  }

  public TokenPair register(RegisterRequest request) {
    if (userRepository.findByEmail(request.getEmail()).isPresent()) {
      throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "User already exists");
    }

    User user = new User();
    user.setEmail(request.getEmail());
    user.setPassword(passwordEncoder.encode(request.getPassword()));
    user.setRefreshTokens(new ArrayList<>());
    userRepository.save(user);

    String accessToken = jwtService.generateAccessToken(user.getId());
    String refreshToken = jwtService.generateRefreshToken(user.getId());

    user.getRefreshTokens().add(refreshToken);
    userRepository.save(user);

    return new TokenPair(accessToken, refreshToken);
  }

  public TokenPair login(LoginRequest request) {
    User user = userRepository.findByEmail(request.getEmail())
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials"));

    if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
      throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials");
    }

    String accessToken = jwtService.generateAccessToken(user.getId());
    String refreshToken = jwtService.generateRefreshToken(user.getId());

    user.getRefreshTokens().add(refreshToken);
    userRepository.save(user);

    return new TokenPair(accessToken, refreshToken);
  }

  public TokenPair refresh(String oldRefreshToken) {
    String userId;
    try {
      userId = jwtService.validateRefreshTokenAndGetUserId(oldRefreshToken);
    } catch (Exception ex) {
      throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid refresh token");
    }

    User user = userRepository.findById(userId)
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "User not found"));

    if (!user.getRefreshTokens().contains(oldRefreshToken)) {
      throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Refresh token revoked");
    }

    String newAccessToken = jwtService.generateAccessToken(userId);
    String newRefreshToken = jwtService.generateRefreshToken(userId);

    user.getRefreshTokens().remove(oldRefreshToken);
    user.getRefreshTokens().add(newRefreshToken);
    userRepository.save(user);

    return new TokenPair(newAccessToken, newRefreshToken);
  }

  public void logout(String userId, String refreshToken) {
    userRepository.findById(userId).ifPresent(user -> {
      user.getRefreshTokens().remove(refreshToken);
      userRepository.save(user);
    });
  }

  public record TokenPair(String accessToken, String refreshToken) {}
}
